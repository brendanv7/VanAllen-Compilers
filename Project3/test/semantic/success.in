{}$

{{{{{{}}}}}}$

{
  print("")
}$

{
  print("int a")
}$

{
  int a
  int b
  boolean c
  c = (a == b)
}$

{
  int a
  {
    string b
  }
  boolean c
}$

{
  boolean d
  d = ("string" != "string")
}$

{
  while true {
    print ((true != (false == true)))
  }
}$

{
  int a
  a = 1
  {
    int b
    {
      int c
      {
        int d
        d = 1
        print ((d == a))
      }
    }
  }
}$
