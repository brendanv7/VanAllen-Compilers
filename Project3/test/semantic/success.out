Compiling program 1...

LEXER -- Lexing program 1...
LEXER -- L_BRACE [ { ] found at (2:1)
LEXER -- R_BRACE [ } ] found at (2:2)
LEXER -- EOP [ $ ] found at (2:3)
LEXER -- Lexing complete for program 1. No errors found.

PARSER -- Parsing program 1...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 1:
<Program>
-<Block>
--{
--<StatementList>
--}
-$


AST for program 1:
<Block>

SEMANTIC -- Analyzing program 1...
SEMANTIC -- Semantic analysis completed successfully with 0 warning(s).

Symbol table for program 1:
ID	Type		Scope	Line	Position

Compilation complete (for now) for program 1.

-------------------------------------------

Compiling program 2...

LEXER -- Lexing program 2...
LEXER -- L_BRACE [ { ] found at (5:1)
LEXER -- L_BRACE [ { ] found at (5:2)
LEXER -- L_BRACE [ { ] found at (5:3)
LEXER -- L_BRACE [ { ] found at (5:4)
LEXER -- L_BRACE [ { ] found at (5:5)
LEXER -- L_BRACE [ { ] found at (5:6)
LEXER -- R_BRACE [ } ] found at (5:7)
LEXER -- R_BRACE [ } ] found at (5:8)
LEXER -- R_BRACE [ } ] found at (5:9)
LEXER -- R_BRACE [ } ] found at (5:10)
LEXER -- R_BRACE [ } ] found at (5:11)
LEXER -- R_BRACE [ } ] found at (5:12)
LEXER -- EOP [ $ ] found at (5:13)
LEXER -- Lexing complete for program 2. No errors found.

PARSER -- Parsing program 2...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 2:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<Block>
-----{
-----<StatementList>
------<Statement>
-------<Block>
--------{
--------<StatementList>
---------<Statement>
----------<Block>
-----------{
-----------<StatementList>
------------<Statement>
-------------<Block>
--------------{
--------------<StatementList>
---------------<Statement>
----------------<Block>
-----------------{
-----------------<StatementList>
-----------------}
---------------<StatementList>
--------------}
------------<StatementList>
-----------}
---------<StatementList>
--------}
------<StatementList>
-----}
---<StatementList>
--}
-$


AST for program 2:
<Block>
-<Block>
--<Block>
---<Block>
----<Block>
-----<Block>

SEMANTIC -- Analyzing program 2...
SEMANTIC -- Semantic analysis completed successfully with 0 warning(s).

Symbol table for program 2:
ID	Type		Scope	Line	Position

Compilation complete (for now) for program 2.

-------------------------------------------

Compiling program 3...

LEXER -- Lexing program 3...
LEXER -- L_BRACE [ { ] found at (8:1)
LEXER -- PRINT [ print ] found at (10:3)
LEXER -- L_PAREN [ ( ] found at (10:8)
LEXER -- QUOTE [ " ] found at (10:9)
LEXER -- QUOTE [ " ] found at (10:10)
LEXER -- R_PAREN [ ) ] found at (10:11)
LEXER -- R_BRACE [ } ] found at (12:1)
LEXER -- EOP [ $ ] found at (12:2)
LEXER -- Lexing complete for program 3. No errors found.

PARSER -- Parsing program 3...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parsePrintStatement()
PARSER -- parseExpr()
PARSER -- parseStringExpr()
PARSER -- parseCharList()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 3:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<PrintStatement>
-----print
-----(
-----<Expr>
------<StringExpr>
-------"
-------<CharList>
-------"
-----)
---<StatementList>
--}
-$


AST for program 3:
<Block>
-<Print Statement>
--<"">

SEMANTIC -- Analyzing program 3...
SEMANTIC -- Semantic analysis completed successfully with 0 warning(s).

Symbol table for program 3:
ID	Type		Scope	Line	Position

Compilation complete (for now) for program 3.

-------------------------------------------

Compiling program 4...

LEXER -- Lexing program 4...
LEXER -- L_BRACE [ { ] found at (15:1)
LEXER -- PRINT [ print ] found at (17:3)
LEXER -- L_PAREN [ ( ] found at (17:8)
LEXER -- QUOTE [ " ] found at (17:9)
LEXER -- CHAR [ i ] found at (17:10)
LEXER -- CHAR [ n ] found at (17:11)
LEXER -- CHAR [ t ] found at (17:12)
LEXER -- SPACE [   ] found at (17:13)
LEXER -- CHAR [ a ] found at (17:14)
LEXER -- QUOTE [ " ] found at (17:15)
LEXER -- R_PAREN [ ) ] found at (17:16)
LEXER -- R_BRACE [ } ] found at (19:1)
LEXER -- EOP [ $ ] found at (19:2)
LEXER -- Lexing complete for program 4. No errors found.

PARSER -- Parsing program 4...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parsePrintStatement()
PARSER -- parseExpr()
PARSER -- parseStringExpr()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 4:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<PrintStatement>
-----print
-----(
-----<Expr>
------<StringExpr>
-------"
-------<CharList>
--------i
--------<CharList>
---------n
---------<CharList>
----------t
----------<CharList>
----------- 
-----------<CharList>
------------a
------------<CharList>
-------"
-----)
---<StatementList>
--}
-$


AST for program 4:
<Block>
-<Print Statement>
--<"int a">

SEMANTIC -- Analyzing program 4...
SEMANTIC -- Semantic analysis completed successfully with 0 warning(s).

Symbol table for program 4:
ID	Type		Scope	Line	Position

Compilation complete (for now) for program 4.

-------------------------------------------

Compiling program 5...

LEXER -- Lexing program 5...
LEXER -- L_BRACE [ { ] found at (22:1)
LEXER -- INT [ int ] found at (24:3)
LEXER -- CHAR [ a ] found at (24:7)
LEXER -- INT [ int ] found at (26:3)
LEXER -- CHAR [ b ] found at (26:7)
LEXER -- BOOL [ boolean ] found at (28:3)
LEXER -- CHAR [ c ] found at (28:11)
LEXER -- CHAR [ c ] found at (30:3)
LEXER -- ASSIGN_OP [ = ] found at (30:5)
LEXER -- L_PAREN [ ( ] found at (30:7)
LEXER -- CHAR [ a ] found at (30:8)
LEXER -- BOOL_OP [ == ] found at (30:10)
LEXER -- CHAR [ b ] found at (30:13)
LEXER -- R_PAREN [ ) ] found at (30:14)
LEXER -- R_BRACE [ } ] found at (32:1)
LEXER -- EOP [ $ ] found at (32:2)
LEXER -- Lexing complete for program 5. No errors found.

PARSER -- Parsing program 5...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseAssignmentStatement()
PARSER -- parseId()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseExpr()
PARSER -- parseId()
PARSER -- parseExpr()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 5:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<VarDecl>
-----int
-----<Id>
------a
---<StatementList>
----<Statement>
-----<VarDecl>
------int
------<Id>
-------b
----<StatementList>
-----<Statement>
------<VarDecl>
-------boolean
-------<Id>
--------c
-----<StatementList>
------<Statement>
-------<AssignmentStatement>
--------<Id>
---------c
--------=
--------<Expr>
---------<BooleanExpr>
----------(
----------<Expr>
-----------<Id>
------------a
----------==
----------<Expr>
-----------<Id>
------------b
----------)
------<StatementList>
--}
-$


AST for program 5:
<Block>
-<Variable Declaration>
--<int>
--<a>
-<Variable Declaration>
--<int>
--<b>
-<Variable Declaration>
--<boolean>
--<c>
-<Assignment Statement>
--<c>
--<==>
---<a>
---<b>

SEMANTIC -- Analyzing program 5...
SEMANTIC -- WARNING: Identifier <a> is declared, but never initialized, at (24:7)
SEMANTIC -- WARNING: Identifier <b> is declared, but never initialized, at (26:7)
SEMANTIC -- WARNING: Identifier <c> initialized, but never used, at (28:11)
SEMANTIC -- Semantic analysis completed successfully with 3 warning(s).

Symbol table for program 5:
ID	Type		Scope	Line	Position
<a>	<int>		0	24	7
<b>	<int>		0	26	7
<c>	<boolean>	0	28	11

Compilation complete (for now) for program 5.

-------------------------------------------

Compiling program 6...

LEXER -- Lexing program 6...
LEXER -- L_BRACE [ { ] found at (35:1)
LEXER -- INT [ int ] found at (37:3)
LEXER -- CHAR [ a ] found at (37:7)
LEXER -- L_BRACE [ { ] found at (39:3)
LEXER -- STRING [ string ] found at (41:5)
LEXER -- CHAR [ b ] found at (41:12)
LEXER -- R_BRACE [ } ] found at (43:3)
LEXER -- BOOL [ boolean ] found at (45:3)
LEXER -- CHAR [ c ] found at (45:11)
LEXER -- R_BRACE [ } ] found at (47:1)
LEXER -- EOP [ $ ] found at (47:2)
LEXER -- Lexing complete for program 6. No errors found.

PARSER -- Parsing program 6...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 6:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<VarDecl>
-----int
-----<Id>
------a
---<StatementList>
----<Statement>
-----<Block>
------{
------<StatementList>
-------<Statement>
--------<VarDecl>
---------string
---------<Id>
----------b
-------<StatementList>
------}
----<StatementList>
-----<Statement>
------<VarDecl>
-------boolean
-------<Id>
--------c
-----<StatementList>
--}
-$


AST for program 6:
<Block>
-<Variable Declaration>
--<int>
--<a>
-<Block>
--<Variable Declaration>
---<string>
---<b>
-<Variable Declaration>
--<boolean>
--<c>

SEMANTIC -- Analyzing program 6...
SEMANTIC -- WARNING: Identifier <a> is declared, but never initialized, at (37:7)
SEMANTIC -- WARNING: Identifier <b> is declared, but never initialized, at (41:12)
SEMANTIC -- WARNING: Identifier <c> is declared, but never initialized, at (45:11)
SEMANTIC -- Semantic analysis completed successfully with 3 warning(s).

Symbol table for program 6:
ID	Type		Scope	Line	Position
<a>	<int>		0	37	7
<b>	<string>	1	41	12
<c>	<boolean>	0	45	11

Compilation complete (for now) for program 6.

-------------------------------------------

Compiling program 7...

LEXER -- Lexing program 7...
LEXER -- L_BRACE [ { ] found at (50:1)
LEXER -- BOOL [ boolean ] found at (52:3)
LEXER -- CHAR [ d ] found at (52:11)
LEXER -- CHAR [ d ] found at (54:3)
LEXER -- ASSIGN_OP [ = ] found at (54:5)
LEXER -- L_PAREN [ ( ] found at (54:7)
LEXER -- QUOTE [ " ] found at (54:8)
LEXER -- CHAR [ s ] found at (54:9)
LEXER -- CHAR [ t ] found at (54:10)
LEXER -- CHAR [ r ] found at (54:11)
LEXER -- CHAR [ i ] found at (54:12)
LEXER -- CHAR [ n ] found at (54:13)
LEXER -- CHAR [ g ] found at (54:14)
LEXER -- QUOTE [ " ] found at (54:15)
LEXER -- BOOL_OP [ != ] found at (54:17)
LEXER -- QUOTE [ " ] found at (54:20)
LEXER -- CHAR [ s ] found at (54:21)
LEXER -- CHAR [ t ] found at (54:22)
LEXER -- CHAR [ r ] found at (54:23)
LEXER -- CHAR [ i ] found at (54:24)
LEXER -- CHAR [ n ] found at (54:25)
LEXER -- CHAR [ g ] found at (54:26)
LEXER -- QUOTE [ " ] found at (54:27)
LEXER -- R_PAREN [ ) ] found at (54:28)
LEXER -- R_BRACE [ } ] found at (56:1)
LEXER -- EOP [ $ ] found at (56:2)
LEXER -- Lexing complete for program 7. No errors found.

PARSER -- Parsing program 7...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseAssignmentStatement()
PARSER -- parseId()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseExpr()
PARSER -- parseStringExpr()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseExpr()
PARSER -- parseStringExpr()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseCharList()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 7:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<VarDecl>
-----boolean
-----<Id>
------d
---<StatementList>
----<Statement>
-----<AssignmentStatement>
------<Id>
-------d
------=
------<Expr>
-------<BooleanExpr>
--------(
--------<Expr>
---------<StringExpr>
----------"
----------<CharList>
-----------s
-----------<CharList>
------------t
------------<CharList>
-------------r
-------------<CharList>
--------------i
--------------<CharList>
---------------n
---------------<CharList>
----------------g
----------------<CharList>
----------"
--------!=
--------<Expr>
---------<StringExpr>
----------"
----------<CharList>
-----------s
-----------<CharList>
------------t
------------<CharList>
-------------r
-------------<CharList>
--------------i
--------------<CharList>
---------------n
---------------<CharList>
----------------g
----------------<CharList>
----------"
--------)
----<StatementList>
--}
-$


AST for program 7:
<Block>
-<Variable Declaration>
--<boolean>
--<d>
-<Assignment Statement>
--<d>
--<!=>
---<"string">
---<"string">

SEMANTIC -- Analyzing program 7...
SEMANTIC -- WARNING: Identifier <d> initialized, but never used, at (52:11)
SEMANTIC -- Semantic analysis completed successfully with 1 warning(s).

Symbol table for program 7:
ID	Type		Scope	Line	Position
<d>	<boolean>	0	52	11

Compilation complete (for now) for program 7.

-------------------------------------------

Compiling program 8...

LEXER -- Lexing program 8...
LEXER -- L_BRACE [ { ] found at (59:1)
LEXER -- WHILE [ while ] found at (61:3)
LEXER -- BOOL_VAL [ true ] found at (61:9)
LEXER -- L_BRACE [ { ] found at (61:14)
LEXER -- PRINT [ print ] found at (63:5)
LEXER -- L_PAREN [ ( ] found at (63:11)
LEXER -- L_PAREN [ ( ] found at (63:12)
LEXER -- BOOL_VAL [ true ] found at (63:13)
LEXER -- BOOL_OP [ != ] found at (63:18)
LEXER -- L_PAREN [ ( ] found at (63:21)
LEXER -- BOOL_VAL [ false ] found at (63:22)
LEXER -- BOOL_OP [ == ] found at (63:28)
LEXER -- BOOL_VAL [ true ] found at (63:31)
LEXER -- R_PAREN [ ) ] found at (63:35)
LEXER -- R_PAREN [ ) ] found at (63:36)
LEXER -- R_PAREN [ ) ] found at (63:37)
LEXER -- R_BRACE [ } ] found at (65:3)
LEXER -- R_BRACE [ } ] found at (67:1)
LEXER -- EOP [ $ ] found at (67:2)
LEXER -- Lexing complete for program 8. No errors found.

PARSER -- Parsing program 8...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseWhileStatement()
PARSER -- parseBooleanExpr()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parsePrintStatement()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 8:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<WhileStatement>
-----while
-----<BooleanExpr>
------true
-----<Block>
------{
------<StatementList>
-------<Statement>
--------<PrintStatement>
---------print
---------(
---------<Expr>
----------<BooleanExpr>
-----------(
-----------<Expr>
------------<BooleanExpr>
-------------true
-----------!=
-----------<Expr>
------------<BooleanExpr>
-------------(
-------------<Expr>
--------------<BooleanExpr>
---------------false
-------------==
-------------<Expr>
--------------<BooleanExpr>
---------------true
-------------)
-----------)
---------)
-------<StatementList>
------}
---<StatementList>
--}
-$


AST for program 8:
<Block>
-<While Statement>
--<true>
--<Block>
---<Print Statement>
----<!=>
-----<true>
-----<==>
------<false>
------<true>

SEMANTIC -- Analyzing program 8...
SEMANTIC -- Semantic analysis completed successfully with 0 warning(s).

Symbol table for program 8:
ID	Type		Scope	Line	Position

Compilation complete (for now) for program 8.

-------------------------------------------

Compiling program 9...

LEXER -- Lexing program 9...
LEXER -- L_BRACE [ { ] found at (70:1)
LEXER -- INT [ int ] found at (72:3)
LEXER -- CHAR [ a ] found at (72:7)
LEXER -- CHAR [ a ] found at (74:3)
LEXER -- ASSIGN_OP [ = ] found at (74:5)
LEXER -- DIGIT [ 1 ] found at (74:7)
LEXER -- L_BRACE [ { ] found at (76:3)
LEXER -- INT [ int ] found at (78:5)
LEXER -- CHAR [ b ] found at (78:9)
LEXER -- L_BRACE [ { ] found at (80:5)
LEXER -- INT [ int ] found at (82:7)
LEXER -- CHAR [ c ] found at (82:11)
LEXER -- L_BRACE [ { ] found at (84:7)
LEXER -- INT [ int ] found at (86:9)
LEXER -- CHAR [ d ] found at (86:13)
LEXER -- CHAR [ d ] found at (88:9)
LEXER -- ASSIGN_OP [ = ] found at (88:11)
LEXER -- DIGIT [ 1 ] found at (88:13)
LEXER -- PRINT [ print ] found at (90:9)
LEXER -- L_PAREN [ ( ] found at (90:15)
LEXER -- L_PAREN [ ( ] found at (90:16)
LEXER -- CHAR [ d ] found at (90:17)
LEXER -- BOOL_OP [ == ] found at (90:19)
LEXER -- CHAR [ a ] found at (90:22)
LEXER -- R_PAREN [ ) ] found at (90:23)
LEXER -- R_PAREN [ ) ] found at (90:24)
LEXER -- R_BRACE [ } ] found at (92:7)
LEXER -- R_BRACE [ } ] found at (94:5)
LEXER -- R_BRACE [ } ] found at (96:3)
LEXER -- R_BRACE [ } ] found at (98:1)
LEXER -- EOP [ $ ] found at (98:2)
LEXER -- Lexing complete for program 9. No errors found.

PARSER -- Parsing program 9...
PARSER -- parse()
PARSER -- parseProgram()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseAssignmentStatement()
PARSER -- parseId()
PARSER -- parseExpr()
PARSER -- parseIntExpr()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseBlock()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseVarDecl()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parseAssignmentStatement()
PARSER -- parseId()
PARSER -- parseExpr()
PARSER -- parseIntExpr()
PARSER -- parseStatementList()
PARSER -- parseStatement()
PARSER -- parsePrintStatement()
PARSER -- parseExpr()
PARSER -- parseBooleanExpr()
PARSER -- parseExpr()
PARSER -- parseId()
PARSER -- parseExpr()
PARSER -- parseId()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- parseStatementList()
PARSER -- Parse completed successfully.

CST for program 9:
<Program>
-<Block>
--{
--<StatementList>
---<Statement>
----<VarDecl>
-----int
-----<Id>
------a
---<StatementList>
----<Statement>
-----<AssignmentStatement>
------<Id>
-------a
------=
------<Expr>
-------<IntExpr>
--------1
----<StatementList>
-----<Statement>
------<Block>
-------{
-------<StatementList>
--------<Statement>
---------<VarDecl>
----------int
----------<Id>
-----------b
--------<StatementList>
---------<Statement>
----------<Block>
-----------{
-----------<StatementList>
------------<Statement>
-------------<VarDecl>
--------------int
--------------<Id>
---------------c
------------<StatementList>
-------------<Statement>
--------------<Block>
---------------{
---------------<StatementList>
----------------<Statement>
-----------------<VarDecl>
------------------int
------------------<Id>
-------------------d
----------------<StatementList>
-----------------<Statement>
------------------<AssignmentStatement>
-------------------<Id>
--------------------d
-------------------=
-------------------<Expr>
--------------------<IntExpr>
---------------------1
-----------------<StatementList>
------------------<Statement>
-------------------<PrintStatement>
--------------------print
--------------------(
--------------------<Expr>
---------------------<BooleanExpr>
----------------------(
----------------------<Expr>
-----------------------<Id>
------------------------d
----------------------==
----------------------<Expr>
-----------------------<Id>
------------------------a
----------------------)
--------------------)
------------------<StatementList>
---------------}
-------------<StatementList>
-----------}
---------<StatementList>
-------}
-----<StatementList>
--}
-$


AST for program 9:
<Block>
-<Variable Declaration>
--<int>
--<a>
-<Assignment Statement>
--<a>
--<1>
-<Block>
--<Variable Declaration>
---<int>
---<b>
--<Block>
---<Variable Declaration>
----<int>
----<c>
---<Block>
----<Variable Declaration>
-----<int>
-----<d>
----<Assignment Statement>
-----<d>
-----<1>
----<Print Statement>
-----<==>
------<d>
------<a>

SEMANTIC -- Analyzing program 9...
SEMANTIC -- WARNING: Identifier <a> initialized, but never used, at (72:7)
SEMANTIC -- WARNING: Identifier <b> is declared, but never initialized, at (78:9)
SEMANTIC -- WARNING: Identifier <c> is declared, but never initialized, at (82:11)
SEMANTIC -- WARNING: Identifier <d> initialized, but never used, at (86:13)
SEMANTIC -- Semantic analysis completed successfully with 4 warning(s).

Symbol table for program 9:
ID	Type		Scope	Line	Position
<a>	<int>		0	72	7
<b>	<int>		1	78	9
<c>	<int>		2	82	11
<d>	<int>		3	86	13

Compilation complete (for now) for program 9.

-------------------------------------------

